src/Microsoft.AspNetCore.Server.IISIntegration/AuthenticationHandler.cs

```diff
@@ -1,138 +1,99 @@
-                {
-                    case ChallengeBehavior.Automatic:
-                        // If there is a principal already, invoke the forbidden code path
-                        if (User == null)
-                        {
-                            goto case ChallengeBehavior.Unauthorized;
-                        }
-                        else
-                        {
-                            goto case ChallengeBehavior.Forbidden;
-                        }
-                    case ChallengeBehavior.Unauthorized:
-                        HttpContext.Response.StatusCode = 401;
-                        // We would normally set the www-authenticate header here, but IIS does that for us.
-                        break;
-                    case ChallengeBehavior.Forbidden:
-                        HttpContext.Response.StatusCode = 403;
-                        break;
-                }
-                context.Accept();
-            }
+            var tokenHeader = httpContext.Request.Headers[MSAspNetCoreWinAuthToken];
 
-            if (PriorHandler != null)
+            int hexHandle;
+            WindowsPrincipal winPrincipal = null;
+            if (!StringValues.IsNullOrEmpty(tokenHeader)
+                && int.TryParse(tokenHeader, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out hexHandle))
             {
-                return PriorHandler.ChallengeAsync(context);
-            }
+                // Always create the identity if the handle exists, we need to dispose it so it does not leak.
+                var handle = new IntPtr(hexHandle);
+                var winIdentity = new WindowsIdentity(handle);
 
-            return TaskCache.CompletedTask;
-        }
+                // WindowsIdentity just duplicated the handle so we need to close the original.
+                NativeMethods.CloseHandle(handle);
 
-        public void GetDescriptions(DescribeSchemesContext context)
-        {
-            foreach (var description in Options.AuthenticationDescriptions)
-            {
-                context.Accept(description.Items);
+                httpContext.Response.RegisterForDispose(winIdentity);
+                winPrincipal = new WindowsPrincipal(winIdentity);
             }
 
-            if (PriorHandler != null)
-            {
-                PriorHandler.GetDescriptions(context);
-            }
+            return winPrincipal;
         }
 
-        public Task SignInAsync(SignInContext context)
+
+        public Task ChallengeAsync(ChallengeContext context)
         {
-            // Not supported, fall through
-            if (PriorHandler != null)
+            switch (context.Behavior)
             {
-                return PriorHandler.SignInAsync(context);
+                case ChallengeBehavior.Automatic:
+                    // If there is a principal already, invoke the forbidden code path
+                    if (GetUser(context.HttpContext) == null)
```

Comment:
Should be fine, the failure scenarios are primarily for a missing header which will be fast.